src/Orleans/Runtime/RequestContext.cs

```diff
@@ -92,39 +88,43 @@ public static void Set(string key, object value)
         /// <returns>Boolean <c>True</c> if the value was previously in the RequestContext key-value bag and has now been removed, otherwise returns <c>False</c>.</returns>
         public static bool Remove(string key)
         {
-            Dictionary<string, object> values = GetContextData();
+            var values = CallContextData.Value;
 
             if (values == null || values.Count == 0 || !values.ContainsKey(key))
             {
                 return false;
             }
             var newValues = new Dictionary<string, object>(values);
             bool retValue = newValues.Remove(key);
-            SetContextData(newValues);
+            CallContextData.Value = newValues;
             return retValue;
         }
 
         public static void Import(Dictionary<string, object> contextData)
         {
             if (PropagateActivityId)
             {
-                object activityIdObj;
-                if (contextData == null || !contextData.TryGetValue(E2_E_TRACING_ACTIVITY_ID_HEADER, out activityIdObj))
+                object activityIdObj = Guid.Empty;
+                if (contextData?.TryGetValue(E2_E_TRACING_ACTIVITY_ID_HEADER, out activityIdObj) != true)
                 {
-                    activityIdObj = Guid.Empty;
-                }
-
-#if NETSTANDARD
-                ActivityId.Value = (Guid) activityIdObj;
-#else
-                Trace.CorrelationManager.ActivityId = (Guid)activityIdObj;
+                    object legacyActivityIdObj = null;
+#if !NETSTANDARD
+                    if (contextData?.TryGetValue(E2_E_TRACING_LEGACY_ACTIVITY_ID_HEADER, out legacyActivityIdObj) == true)
+                    {
+                        Trace.CorrelationManager.ActivityId = (Guid) legacyActivityIdObj;
+                    }
 #endif
+                }
+                else
+                {
+                    ActivityId.Value = (Guid)activityIdObj;
```

Comment:
I suppose this is enough not to set ActivityId if unused. Do we want to set it to Guid.Empty if `ActivityId.Value` is already set to something else though?
I believe the answer is no, and we should keep the code as-is, but wanted to ask you in case you did not consider that case and you actually have a different opinion about it.